### PR TITLE
Supporting logger factories for image build and pulls

### PR DESCRIPTION
**Pull request by bfosberry**
This change updates how the container and builder write logs from docker, to use a provided LoggerFactory. This currently uses the following assumptions

1) Container loggers are requested using the format $PROJECTNAME_$SERVICENAME_$INSTANCENUM, eg myproject_app_1 as per the compose container name format.
2) Image build loggers are requested using the service name
3) Image pull loggers are requested using the image name

The problem with this is that it's highly likely someone would use the same image and service name, e.g. the redis service uses the image redis. 

Possible solutions:

* Using different logger factories for the different types of loggers
* Extending the logger factory interface to request different types of loggers

thoughts?

### CODE REVIEW

**Comment by bfosberry** [comment]
I'm going to update this with a more dynamic logger factory interface

**Comment by bfosberry** [comment]
@vdemeester Updated this, build failure does not seem related, could you verify and retest if needed? Let me know if you like the unified loggerfactory approach.

thanks!

**Comment by bfosberry** [comment]
thoughts? good to merge?

**Comment by vdemeester** [comment]
@bfosberry Ok, I took time to look at this (sorry for long time, again :sweat:).

In theory I like that, but in order to be *compliant* with `docker-compose` we should do something on the `cli` side to have the simple output there is now. @bfosberry WDYT ?

**Comment by bfosberry** [comment]
Oh I see what you mean, that does look weird. 

For something  CLI side you just mean have a sample factory to keep the output the same?

**Comment by vdemeester** [comment]
@bfosberry yes, providing two loggers, one that display the default output for pull and build (same as docker-compose) and one as these :wink:.

**Comment by bfosberry** [comment]
Picking this up again

**Comment by vdemeester** [comment]
@bfosberry note that *in the end*, the basic idea is also to remove `logrus` from the *core* (https://github.com/docker/libcompose/pull/243 for inspiration)

**Comment by bfosberry** [comment]
:+1:

**Comment by bfosberry** [comment]
@vdemeester I think this is gtg, here is a side by side comparison of master and this branch for the base tool:

Branch:

```
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] Project [merrygoround]: Starting project     
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
6874f9870f5f: Pull complete 
4c876570bd7d: Pull complete 
10fb34ebccea: Pull complete 
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in b2f7e11114f5
 ---> 2e2c23dc6ead
Removing intermediate container b2f7e11114f5
Successfully built 2e2c23dc6ead
INFO[0018] [1/1] [app]: Started                         
INFO[0018] Project [merrygoround]: Project started 
```

branch:

```
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] Project [merrygoround]: Starting project     
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
6874f9870f5f: Pull complete 
4c876570bd7d: Pull complete 
10fb34ebccea: Pull complete 
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in b94eeed630f5
 ---> 11d5c023a302
Removing intermediate container b94eeed630f5
Successfully built 11d5c023a302
INFO[0021] [1/1] [app]: Started                         
INFO[0021] Project [merrygoround]: Project started
```

**Comment by bfosberry** [comment]
Yeah, looks like the log rendering still looks abnormal atm. Will fix

**Comment by bfosberry** [comment]
Fixed it here is an updated sdie-by-side comparison

branch:

```
bfosberry@slothula:~/workspace/merrygoround$ $GOPATH/src/github.com/docker/libcompose/bundles/libcompose-cli up
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pulling fs layer
6874f9870f5f: Pulling fs layer
4c876570bd7d: Pulling fs layer
10fb34ebccea: Pulling fs layer
10fb34ebccea: Waiting
4c876570bd7d: Verifying Checksum
4c876570bd7d: Download complete
6874f9870f5f: Verifying Checksum
6874f9870f5f: Download complete
10fb34ebccea: Verifying Checksum
10fb34ebccea: Download complete
5ba4f30e5bea: Verifying Checksum
5ba4f30e5bea: Download complete
5ba4f30e5bea: Pull complete
6874f9870f5f: Pull complete
4c876570bd7d: Pull complete
10fb34ebccea: Pull complete
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in d23ef4939444
 ---> fb7d65b4b138
Removing intermediate container d23ef4939444
Successfully built fb7d65b4b138
INFO[0023] [1/1] [app]: Started     
```

master:

```
bfosberry@slothula:~/workspace/merrygoround$ $GOPATH/src/github.com/docker/libcompose/bundles/libcompose-cli up
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
6874f9870f5f: Pull complete 
4c876570bd7d: Pull complete 
10fb34ebccea: Pull complete 
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in 24fc45d30df0
 ---> b541c342aa0b
Removing intermediate container 24fc45d30df0
Successfully built b541c342aa0b
INFO[0022] [1/1] [app]: Started     
```

I think the only difference is that the pulling/verifying/downloading/pull complete messages dont get squashed. Unsure why that is.

**Comment by vdemeester** [comment]
@bfosberry a unit-test is failing (panic :sweat:)

**Comment by bfosberry** [comment]
Yeah, I just need to add a nil check where used, auto-create it, it just
set it with tests.

What do you think of the log differences though?

**Comment by vdemeester** [comment]
@bfosberry there is a little bit too much output :sweat:. I feel like we should change that `outFd, isTerminalOut := term.GetFdInfo(buildBuff)` so that the build logger would return the *inner* writer used (in the case of the cli `os.Stderr` or `os.Stdout`).

**Comment by bfosberry** [comment]
New comparison:

this change:
```
bfosberry@slothra:~/workspace/merrygoround$ $GOPATH/src/github.com/docker/libcompose/bundles/libcompose-cli up
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
6874f9870f5f: Pull complete 
4c876570bd7d: Pull complete 
10fb34ebccea: Pull complete 
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in 889397e841fa
 ---> 7f2053c03438
Removing intermediate container 889397e841fa
Successfully built 7f2053c03438
INFO[0019] [1/1] [app]: Started 
```

master branch: 

```
bfosberry@slothra:~/workspace/merrygoround$ $GOPATH/src/github.com/docker/libcompose/bundles/libcompose-master up
WARN[0000] Note: This is an experimental alternate implementation of the Compose CLI (https://github.com/docker/compose) 
INFO[0000] [0/1] [app]: Starting                        
WARN[0000] Error while reading .dockerignore (.dockerignore) : open .dockerignore: no such file or directory 
INFO[0000] Building merrygoround_app...                 
Sending build context to Docker daemon 20.55 MB
Step 1 : FROM ubuntu
latest: Pulling from library/ubuntu
5ba4f30e5bea: Pull complete 
6874f9870f5f: Pull complete 
4c876570bd7d: Pull complete 
10fb34ebccea: Pull complete 
Digest: sha256:f1b592e2de671105255a0c0b7b2f71a92b829403e8fc845e3482667ecc301780
Status: Downloaded newer image for ubuntu:latest
 ---> 12543ced0f6f
Step 2 : RUN touch /codeship-docker-push-example
 ---> Running in 5297480822e2
 ---> 4b5daa8ba1b8
Removing intermediate container 5297480822e2
Successfully built 4b5daa8ba1b8
INFO[0019] [1/1] [app]: Started
```

**Comment by vdemeester** [comment]
@bfosberry looks good yeah, just some `golint` errors though :angel: 

```
20:54:51 Errors from golint:
20:54:51 logger/null.go:35:1: receiver name c should be consistent with previous receiver name n for NullLogger
20:54:51 logger/null.go:40:1: receiver name c should be consistent with previous receiver name n for NullLogger
20:54:51 logger/raw_logger.go:41:1: receiver name c should be consistent with previous receiver name n for RawLogger
20:54:51 logger/raw_logger.go:46:1: receiver name c should be consistent with previous receiver name n for RawLogger
```

**Comment by bfosberry** [comment]
Oh derp, will fix

**Comment by bfosberry** [comment]
I get a panic locally, will keep looking into it here

**Comment by bfosberry** [comment]
I think it looks good now!

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,7 @@ LIBCOMPOSE_ENVS := \
 	-e TESTVERBOSE
 
 # (default to no bind mount if DOCKER_HOST is set)
-BIND_DIR := $(if $(DOCKER_HOST),,bundles)
```
Comment (vdemeester) [inline Makefile]: Why this change ? should not be needed (and is not related)

File: cli/logger/color_logger.go [hunk 2]
```diff
@@ -29,8 +30,23 @@ func NewColorLoggerFactory() *ColorLoggerFactory {
 	}
 }
 
-// Create implements logger.Factory.Create.
-func (c *ColorLoggerFactory) Create(name string) logger.Logger {
+// CreateContainerLogger implements logger.Factory.CreateContainerLogger.
+func (c *ColorLoggerFactory) CreateContainerLogger(name string) logger.Logger {
+	return c.create(name)
```
Comment (vdemeester) [inline cli/logger/color_logger.go]: why returning `c.create` and not only `c` ?

File: docker/builder/builder.go [hunk 3]
```diff
@@ -52,9 +54,30 @@ func (d *DaemonBuilder) Build(ctx context.Context, imageName string) error {
 		return err
 	}
 	defer buildCtx.Close()
+	var l logger.Logger
+	if d.LoggerFactory != nil {
+		l = d.LoggerFactory.CreateBuildLogger(imageName)
+	}
 
 	var progBuff io.Writer = os.Stdout
```
Comment (vdemeester) [inline docker/builder/builder.go]: I wonder about that : Should we make the assumption that the loggerFactory is not nil, and/or have a default one that just *eats* the output ? This would remove the need to check for `nil` in multiple places here.

Comment (vdemeester) [inline docker/builder/builder.go]: /cc @joshwget

File: Makefile (hunk 1 above)
Comment (bfosberry) [inline Makefile]: Damnit that was testing, will remove

File: cli/logger/color_logger.go (hunk 2 above)
Comment (bfosberry) [inline cli/logger/color_logger.go]: ColorLoggerFactory does not also implement the LoggerInterface, its doing some work to wrap the factory via a logger interface

File: docker/builder/builder.go (hunk 3 above)
Comment (bfosberry) [inline docker/builder/builder.go]: Hmmm, we could try and enforce a logger factory. Maybe if we move some components to constructors where we can verify the fields?

Comment (joshwget) [inline docker/builder/builder.go]: I haven't checked out this PR too much yet, but enforcing a logger factory sounds like a good idea to me

Comment (vdemeester) [inline docker/builder/builder.go]: Same though as @joshwget

Comment (bfosberry) [inline docker/builder/builder.go]: Updated, let me know what you think

**Comment by bfosberry** [comment]
@vdemeester, @joshwget let me know if these changes look good

thanks!

**Comment by vdemeester** [comment]
LGTM 🐮
/cc @joshwget

**Comment by joshwget** [comment]
Should we move `ColorLogger` and `ColorLoggerFactory` to the top-level `logger` package? These are useful even when not making use of the `cli` package, so it seems a little misplaced as a subpackage there.

**Comment by joshwget** [comment]
Actually, we can discuss moving `ColorLogger` and `ColorLoggerFactory` separately. That's not really relevant to this PR.

**Comment by joshwget** [comment]
@bfosberry Would you mind squashing the commits?

**Comment by vdemeester** [comment]
ping @bfosberry, this needs a rebase too :angel: 
then it should be ready to be merged :wink:

**Comment by bfosberry** [comment]
Oh yes, totally. Can squash, should have tume this week.

thanks guys!

**Comment by bfosberry** [comment]
Hey, thanks for the patience on this, squashed, fixed rebase issues, looks good!

**Comment by vdemeester** [comment]
Hum failure are weird, I'll see if this happens all the time or just a *bad* timeframe for the build :sweat_smile: 

Anyway, ran it @home and it's green :wink: 
LGTM 🐸
